umbraco/Umbraco-CMS
# V14: Document and media collection view endpoints

## Details
- New endpoints for document and media collection views:
  - `GET /umbraco/management/api/v1/collection/document` & 
  - `GET /umbraco/management/api/v1/collection/media`;
  - "collection view" (client) = "list view" (server);
  - New "collection" silo;
  - Gets the collection items by id - in the case of media, we can get all items at root as well by providing `"id" : null`;
- New services for content and media list views and a base class that does all the heavy lifting;
  - There are a few TODOs/FIXMEs for future improvements.
- Changed the mapping of content values (properties) to be more flexible by taking only the selected properties (_instead of all_);
- Adding a new document and media type reference models (following the new mapping principle).

Functionality:
- Get collection items by document/media id (_OR get collection items for media root_) in a paginated manner;
- Results will include both variant and invariant items - we can order by culture (`orderCulture`), but all variants will be part of the result (wish from the client);
- When a `dataTypeId` is not specified, the configured collection view of a document/media type will be checked for its configuration; Its configured properties will be taken into account when returning document/media items - only the properties configured in a collection view will bee part of the document/media results _(`"values"` collection);_
- To specify a `dataTypeId` of a collection view data type, this data type needs to be part of the document/media type properties - it is **NOT** the id of the configured collection view for a document/media type;
- We can only order the results by a property that is part of the collection view configuration OR by the `"name"` field;
- The `filter` parameter was utilized by the list view search bar in the past - its functionality remains:
  - We can pass in a value that is part of the document/media item name or a Guid of a specific document/media item.
---
**Bonus**: Adding a missing `UnauthorizedMissingPermissionAccess` authorization status and returning it from the respective permission services. 😉

## Test
> Setup
- The easiest would be to set up everything on v13 DB from a v13 template project and copy over the DB to this v14 branch;
- Create an **invariant** document type (allow at root - referred to as `Root`) and another **variant** one (allow at root, enable list view, add a list view property on its content - referred to as `Products`) and a child doc type of `Products` - referred to as `Product` with a few properties (`itemName` and `description` /which varies by culture/), and allow it to vary by culture;
- Create a few list view data types and change the columns displayed to verify that when different collection view data types are used, different document/media properties (= `values`) are part of the output;
- Create a new media type with list view enabled - you can stick to the default media type list view or create a custom one (change the properties) and verify that the correct one is picked up when no collection view data type is specified;
  - Add a list view property as well.

> General
- Verify that only list view data type can be specified;
- Verify that to specify a data type, document or media item id is necessary;
- Verify that the specified list view data type has to be part of the document/media type properties;
- Verify that when a list view data type is not specified, the configured list view on the document/media type is considered - _you can verify that by configuring different Include Properties on the List View_;
- Verify that you can only order by properties in the list view configuration;
- Verify that you can filter the result by providing the name or the Guid of concrete document/media item in the `filter` param.
> Document collection endpoint
- Verify that you can get the collection items by a document id;
- Verify that you can sort by `name` and provide a culture - we will get all the variants but the results will be ordered correctly.
> Media collection endpoint
- Verify that not providing any params, returns all items at root;
- Verify that you can get the collection items by a media id.
